manta-trusted-setup/src/util.rs

```diff
@@ -339,54 +338,151 @@ where
     (g1_proj.into_affine(), g2_proj.into_affine())
 }
 
-/// Pair from a [`PairingEngine`]
-type Pair<P> = (
-    <P as PairingEngine>::G1Prepared,
-    <P as PairingEngine>::G2Prepared,
-);
+/// Blake Hasher
+#[derive(Default)]
+pub struct BlakeHasher(pub Blake2b512);
 
-/// Pairing Engine Extension
-pub trait PairingEngineExt: PairingEngine {
-    /// Evaluates the pairing function on `pair`.
+impl Write for BlakeHasher {
     #[inline]
-    fn eval(pair: &Pair<Self>) -> Self::Fqk {
-        Self::product_of_pairings(iter::once(pair))
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.0.update(buf);
+        Ok(buf.len())
     }
 
-    /// Checks if `lhs` and `rhs` evaluate to the same point under the pairing function.
     #[inline]
-    fn has_same(lhs: &Pair<Self>, rhs: &Pair<Self>) -> bool {
-        Self::eval(lhs) == Self::eval(rhs)
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
     }
+}
 
-    /// Checks if `lhs` and `rhs` evaluate to the same point under the pairing function, returning
-    /// `Some` with prepared points if the pairing outcome is the same. This function checks if
-    /// there exists an `r` such that `(r * lhs.0 == rhs.0) && (lhs.1 == r * rhs.1)`.
+impl<P, const N: usize> HashToGroup<P, [u8; N]> for BlakeHasher
+where
+    P: Pairing,
+    P::G2: Sample,
+{
     #[inline]
-    fn same<L1, L2, R1, R2>(lhs: (L1, L2), rhs: (R1, R2)) -> Option<(Pair<Self>, Pair<Self>)>
+    fn hash(&self, challenge: &[u8; N], ratio: (&P::G1, &P::G1)) -> P::G2 {
+        let mut hasher = BlakeHasher::default();
+        hasher.0.update(challenge);
+        ratio.0.serialize(&mut hasher).unwrap();
+        ratio.1.serialize(&mut hasher).unwrap();
+        hash_to_group::<_, (), 64>(into_array_unchecked(hasher.0.finalize()))
+    }
+}
+
+/// KZG Blake Hasher
+pub struct KZGBlakeHasher<C>
+where
+    C: kzg::Configuration + ?Sized,
+{
+    /// Domain Tag Type
+    pub domain_tag: C::DomainTag,
+}
+
+impl<P, const N: usize> HashToGroup<P, [u8; N]> for KZGBlakeHasher<P>
+where
+    P: kzg::Configuration<DomainTag = u8> + ?Sized,
+    P::G2: Sample,
+{
+    #[inline]
+    fn hash(&self, challenge: &[u8; N], ratio: (&P::G1, &P::G1)) -> P::G2 {
+        let mut hasher = BlakeHasher::default();
+        hasher.0.update(&[self.domain_tag]);
+        hasher.0.update(challenge);
+        ratio.0.serialize(&mut hasher).unwrap();
+        ratio.1.serialize(&mut hasher).unwrap();
+        hash_to_group::<_, (), 64>(into_array_unchecked(hasher.0.finalize()))
+    }
+}
+
+/// Hashes `digest` to a group point on affine curve.
```

Comment:
Add some documentation here about the fact that this uses the `digest` as a seed to an RNG and then uses the RNG to sample `G`.